src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs

```diff
@@ -2,219 +2,165 @@
+    private Action<TManager, TEventHandler, TWorkload> _runWorkload;
+    private readonly List<ProviderSpecificWorkload<TWorkload>> _workloads = new();
+    private readonly List<Slot> _managerSlots = new();
 
-    #endregion
+    private readonly object _lock = new();
 
-    protected ParallelOperationManager(Func<T> createNewManager, int parallelLevel, bool sharedHosts)
+    public ParallelOperationManager(Func<TestRuntimeProviderInfo, TManager> createNewManager, int parallelLevel)
     {
-        CreateNewConcurrentManager = createNewManager;
-        SharedHosts = sharedHosts;
-
-        // Update Parallel Level
-        UpdateParallelLevel(parallelLevel);
+        _createNewManager = createNewManager;
+        _maxParallelLevel = parallelLevel;
+        ClearSlots();
     }
 
-    /// <summary>
-    /// Remove and dispose a manager from concurrent list of manager.
-    /// </summary>
-    /// <param name="manager">Manager to remove</param>
-    public void RemoveManager(T manager)
+    private void ClearSlots()
     {
-        _concurrentManagerHandlerMap.Remove(manager);
+        lock (_lock)
+        {
+            _managerSlots.Clear();
+            _managerSlots.AddRange(Enumerable.Range(0, _maxParallelLevel).Select(_ => new Slot()));
+        }
     }
 
-    /// <summary>
-    /// Add a manager in concurrent list of manager.
-    /// </summary>
-    /// <param name="manager">Manager to add</param>
-    /// <param name="handler">eventHandler of the manager</param>
-    public void AddManager(T manager, TU handler)
+    public void StartWork(
+        List<ProviderSpecificWorkload<TWorkload>> workloads!!,
+        TEventHandler eventHandler!!,
+        Func<TEventHandler, TManager, TEventHandler> getEventHandler!!,
+        Action<TManager, TEventHandler, TWorkload> runWorkload!!)
     {
-        _concurrentManagerHandlerMap.Add(manager, handler);
-    }
+        _eventHandler = eventHandler;
+        _getEventHandler = getEventHandler;
+        _runWorkload = runWorkload;
 
-    /// <summary>
-    /// Update event handler for the manager.
-    /// If it is a new manager, add this.
-    /// </summary>
-    /// <param name="manager">Manager to update</param>
-    /// <param name="handler">event handler to update for manager</param>
-    public void UpdateHandlerForManager(T manager, TU handler)
-    {
-        if (_concurrentManagerHandlerMap.ContainsKey(manager))
-        {
-            _concurrentManagerHandlerMap[manager] = handler;
-        }
-        else
+        _workloads.AddRange(workloads);
+
+        lock (_lock)
         {
-            AddManager(manager, handler);
+            ClearSlots();
+            RunWorkInParallel();
         }
     }
 
-    /// <summary>
-    /// Get the event handler associated with the manager.
-    /// </summary>
-    /// <param name="manager">Manager</param>
-    public TU GetHandlerForGivenManager(T manager)
+    private bool RunWorkInParallel()
```

Comment:
This does not do anything in parallel, all the workload we schedule is offloaded to separate tasks in the callback. I did not want to change that, but this is the correct place to do that offloading. Not each manager.

Reply:
Added